pandas/_libs/lib.pyx

```diff
@@ -1364,7 +1372,10 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
 
     elif PyDateTime_Check(val):
         if is_datetime_array(values):
-            return "datetime"
+            if is_datetimetz_array(values):
```

Comment:
this is likely enough

Reply:
I will very simply delete the part above and test locally. I'll check again which paths the tests function take. Perhaps it is not useful or perhaps there is a better way.

Reply:
So @jreback I executed my plan and one of the tests failed (see code of failed test below) after I removed the part which is not in the correct place (<code>elif hasattr(value, 'dtype') and hasattr(value.dtype, 'tz'):...</code>).

I followed the path taken during the test. As you can see by the code and my comments below (look for comments with lots of hyphens) it goes through _try_infer_map (see code below) where it is not recognized as "datetimetz" but as "datetime64" (s.dtype.kind is "M" which in _TYPE_MAP translates to "datetime64").

So the kind and base attribute of s.dtype don't help identify our Series as "datetimetz" with try_infer_map (see code below) and since it's a mapping and not say regex the name attribute of s.dtype which is "datetime64[ns, US/Pacific]" won't match (we'd have to include all timezones in the mapper which does not seem like the right solution).

Between this and the fact that afterwards the dtype is supposed to be returned this brings me to the conclusion that the check for DatetimeTZdtype must be right before try_infer_map as it was during a [previous commit](https://github.com/pandas-dev/pandas/pull/30492/commits/da52caf536e3012d786332b79011bee0eed82151#diff-a23dec10f398de8ed7358a4b4eca76c9R1300-R1301) (or we would have to do some surgery and check no other inferences break).


**_failed_test_**
```python
import pandas as pd
from pandas import _lib as lib
from pandas import Series

# pandas/tests/series/test_constructors.py
# TestSeriesConstructors.test_constructor_with_datetime_tz
# inference
s = Series([pd.Timestamp("2013-01-01 13:00:00-0800", tz="US/Pacific"),
            pd.Timestamp("2013-01-02 14:00:00-0800", tz="US/Pacific")])
# Assertion Error 'datetime64' == 'datetimetz'
assert lib.infer_dtype(s, skipna=True) == "datetimetz"
```


**_infer_dtype_**
```python
if util.is_array(value):
    values = value

elif hasattr(value, 'dtype'):  # <------------ GOES HERE
    # this will handle ndarray-like
    # e.g. categoricals
    try:
        values = getattr(value, '_values', getattr(value, 'values', value))
    except TypeError:  # <------------ GOES HERE
        # This gets hit if we have an EA, since cython expects `values`
        #  to be an ndarray
        value = _try_infer_map(value) # -----SEE try_infer_map code below----
        if value is not None:
            return value

        # its ndarray like but we can't handle
        raise ValueError(f"cannot infer type for {type(value)}")

else:
    if not isinstance(value, list):
        value = list(value)
    from pandas.core.dtypes.cast import (
        construct_1d_object_array_from_listlike)
    values = construct_1d_object_array_from_listlike(value)
```
**_try_infer_map_**

```cython
_TYPE_MAP = {'categorical': 'categorical', 'category': 'categorical',
             'int8': 'integer', 'int16': 'integer', 'int32': 'integer',
             'int64': 'integer', 'i': 'integer', 'uint8': 'integer',
             'uint16': 'integer', 'uint32': 'integer', 'uint64': 'integer',
             'u': 'integer', 'float32': 'floating', 'float64': 'floating',
             'f': 'floating', 'complex64': 'complex', 'complex128': 'complex',
             'c': 'complex', 'string': 'string', 'S': 'bytes', 'U': 'string', 
             'bool': 'boolean', 'b': 'boolean', 'datetime64[ns]': 'datetime64',
             'M': 'datetime64', 'timedelta64[ns]': 'timedelta64',
             'm': 'timedelta64', 'interval': 'interval'}

# NOT COPIED: types which only exist on certain platform
# (it's for float16, flat128 and complex256)

cdef _try_infer_map(v):
    """
    If its in our map, just return the dtype.
    """
    cdef:
        object attr, val
    for attr in ['name', 'kind', 'base']:
        # -----------------------------------
        # for our example:
        # name datetime64[ns, US/Pacific]
        # kind M
        # base dtype('<M8[ns]')
        # note: base won't be called bc kind "M" in is _TYPE_MAP
        # -----------------------------------
        val = getattr(v.dtype, attr)
        if val in _TYPE_MAP:
            return _TYPE_MAP[val]
    return None
```

Reply:
can we do this without doing multiple passes through the data?

Reply:
@jbrockmendel cython.datetime does not seem to provide a way to distinguish between datetime with and without timezone however this is something we can do I think (I tested it locally and it works): in DatetimeValidator we can check the absence of tzinfo.

```cython
cdef class DatetimeValidator(TemporalValidator):
    cdef bint is_value_typed(self, object value) except -1:
        return PyDateTime_Check(value) and (value.tzinfo is None) # <-----
```

And then we change the part with the repetition to:

```python
    elif PyDateTime_Check(val):
        if is_datetime_array(values):
            return "datetime"
        elif is_datetimetz_array(values):
            return "datetimetz"    
```